ambry-tools/src/main/java/com/github/ambry/tools/perf/serverperf/ServerPerfNetworkQueue.java

```diff
@@ -0,0 +1,256 @@
+   *  4. Removes all the collected responses from pending correlation id maps and adds to {@link #responseInfos}
+   *
+   *
+   *  If shutdown is triggered , waits for all responses to be processed and shuts down  {@link #executorService}
+   */
+  @Override
+  public void run() {
+    while (!isShutDown || !requestInfos.isEmpty() || !pendingCorrelationIdToStartTimeMs.isEmpty()) {
+      List<ResponseInfo> responseInfos = new ArrayList<>();
+
+      Iterator<Map.Entry<Integer, Long>> pendingRequestIterator =
+          pendingCorrelationIdToStartTimeMs.entrySet().iterator();
+
+      while (pendingRequestIterator.hasNext()) {
+        Map.Entry<Integer, Long> entry = pendingRequestIterator.next();
+        int correlationId = entry.getKey();
+        long startTimeMs = entry.getValue();
+
+        long currentTimeInMs = time.milliseconds();
+
+        if (startTimeMs + operationsTimeOutMs < currentTimeInMs) {
+
+          responseInfos.add(new ResponseInfo(pendingCorrelationIdToRequestInfo.get(correlationId),
+              NetworkClientErrorCode.TimeoutError, null));
+
+          responseInfos.addAll(networkClients.get(pendingCorrelationIdToNetworkClientIdx.get(correlationId))
+              .sendAndPoll(Collections.emptyList(), new HashSet<>(Collections.singletonList(correlationId)),
+                  pollTimeout));
+          pendingRequestIterator.remove();
+        } else {
+          break;
+        }
+      }
+
+      networkClients.forEach(networkClient -> {
+        responseInfos.addAll(networkClient.sendAndPoll(Collections.emptyList(), new HashSet<>(), pollTimeout));
+      });
+
+      if (!requestInfos.isEmpty()) {
+        RequestInfo requestInfo = requestInfos.peek();
+
+        pendingCorrelationIdToStartTimeMs.put(requestInfo.getRequest().getCorrelationId(), time.milliseconds());
+        pendingCorrelationIdToNetworkClientIdx.put(requestInfo.getRequest().getCorrelationId(), clientIndex);
+        pendingCorrelationIdToRequestInfo.put(requestInfo.getRequest().getCorrelationId(), requestInfo);
+        requestInfos.remove();
+
+        List<ResponseInfo> responses = networkClients.get(clientIndex)
+            .sendAndPoll(Collections.singletonList(requestInfo), new HashSet<>(), pollTimeout);
+        clientIndex++;
+        clientIndex = clientIndex % networkClients.size();
+        responseInfos.addAll(responses);
+      }
+
+      this.responseInfos.addAll(responseInfos);
+
+      responseInfos.forEach(responseInfo -> {
+        pendingCorrelationIdToStartTimeMs.remove(responseInfo.getRequestInfo().getRequest().getCorrelationId());
+        pendingCorrelationIdToNetworkClientIdx.remove(responseInfo.getRequestInfo().getRequest().getCorrelationId());
+        pendingCorrelationIdToRequestInfo.remove(responseInfo.getRequestInfo().getRequest().getCorrelationId());
+      });
+    }
+
+    try {
+      maxParallelRequest.release(1);
+      maxParallelRequest.acquire(maxParallelism + 1);
```

Comment:
I don't think this is going to work. You want to block until all the responses are processed by the executor. But it's not guaranteed that there will be maxParallelism responses being processed when you call release in line 237.

To make it accurate, we should jsut create another countdown latch, when all response are processed in the poll method, count it down.

Reply:
Added release in submit too. 
First release triggers the the thread that is trying to submit and waiting and shutdown occurs. In case of shutdown this thread also releases a token.

Then we are waiting for all released tokens to be available. Initially there are maxParallelism tokens. When all responses are processed in poll method, total released tokens will be maxParallelism +1 , Then the we stop waiting.